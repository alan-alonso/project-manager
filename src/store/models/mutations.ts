import { ProjectStatus } from "@/models/project";

export interface ChangeProjectStatusPayload {
  projectId: string;
  newStatus: ProjectStatus;
}
