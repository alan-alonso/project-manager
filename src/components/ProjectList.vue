<template>
  <v-container
    :class="projectClassObj"
    class="lighten-1 mt-1"
    fluid
    @dragover.prevent="onDragOver"
    @drop.prevent="onDrop"
  >
    <h2 class="white--text text-center mb-2">{{ listType }} PROJECTS</h2>
    <Project
      v-for="project in getProjectList()"
      :key="project.id"
      :project="project"
      class="mb-2"
    />
  </v-container>
</template>

<script lang="ts">
import { Vue, Component, Prop } from "vue-property-decorator";
import { ProjectStatus, Project } from "../models/project";
import { Getter, Action } from "vuex-class";
import ProjectCmp from "@/components/Project.vue";
import { ChangeProjectStatusPayload } from "../store/models/mutations";

@Component({
  name: "ProjectList",
  components: {
    Project: ProjectCmp
  }
})
export default class ProjectList extends Vue {
  @Prop({ required: true })
  listType!: ProjectStatus;

  projectClassObj = {
    pink: this.listType === ProjectStatus.ACTIVE,
    indigo: this.listType === ProjectStatus.FINISHED
  };

  @Getter("activeProjects")
  activeProjects!: Project[];

  @Getter("finishedProjects")
  finishedProjects!: Project[];

  @Action("changeProjectStatus")
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  changeProjectStatus: any;

  getProjectList() {
    switch (this.listType) {
      case ProjectStatus.ACTIVE:
        return this.activeProjects;
      case ProjectStatus.FINISHED:
        return this.finishedProjects;
      default:
        return [];
    }
  }

  // eslint-disable-next-line @typescript-eslint/no-empty-function
  onDragOver() {}

  onDrop(event: DragEvent) {
    if (event.dataTransfer && event.dataTransfer.types[0] === "text/plain") {
      const projectId = event.dataTransfer.getData("text/plain");
      const payload: ChangeProjectStatusPayload = {
        projectId,
        newStatus: this.listType
      };
      this.changeProjectStatus(payload);
    }
  }
}
</script>
