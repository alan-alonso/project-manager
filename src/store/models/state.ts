import { Project } from "@/models/project";

export interface StateModel {
  projects: Project[];
}
