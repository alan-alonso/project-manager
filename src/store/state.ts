import { StateModel } from "@/store/models/state";

const state: StateModel = {
  projects: []
};

export default state;
