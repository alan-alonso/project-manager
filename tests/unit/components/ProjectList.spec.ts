import Vue from "vue";
import Vuex, { Store } from "vuex";
import Vuetify from "vuetify";

Vue.use(Vuetify);

import { createLocalVue, shallowMount, mount } from "@vue/test-utils";
import ProjectList from "@/components/ProjectList.vue";
import ProjectCmp from "@/components/Project.vue";
import { ProjectStatus, Project } from "@/models/project";
import { StateModel } from "@/store/models/state";

const localVue = createLocalVue();
localVue.use(Vuex);

describe("ProjectList.vue", () => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let vuetify: any;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let actions: any;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let getters: any;
  let store: Store<StateModel>;

  beforeEach(() => {
    vuetify = new Vuetify();
    getters = {
      projects: jest.fn(),
      activeProjects: jest.fn(),
      finishedProjects: jest.fn()
    };
    actions = {
      changeProjectStatus: jest.fn()
    };
    store = new Vuex.Store({
      actions,
      getters
    });
  });

  test("should create active project list with pink background color", () => {
    // act
    const wrapper = mount(ProjectList, {
      localVue,
      vuetify,
      store,
      propsData: {
        listType: ProjectStatus.ACTIVE
      }
    });

    // assert
    expect(wrapper.element.classList).toContain("pink");
  });

  test("should create finished project list with indigo background color", () => {
    // act
    const wrapper = mount(ProjectList, {
      localVue,
      vuetify,
      store,
      propsData: {
        listType: ProjectStatus.FINISHED
      }
    });

    // assert
    expect(wrapper.element.classList).toContain("indigo");
  });

  test("should render project list with store getter provided for projects", () => {
    // arrange
    store.hotUpdate({
      getters: {
        activeProjects: (): Project[] => [
          {
            id: "1",
            title: "Test Title",
            description: "Test Description",
            people: 1,
            status: ProjectStatus.ACTIVE
          }
        ]
      }
    });

    // act
    const wrapper = shallowMount(ProjectList, {
      localVue,
      vuetify,
      store,
      propsData: {
        listType: ProjectStatus.ACTIVE
      }
    });

    // assert
    const Projects = wrapper.findAll(ProjectCmp);
    expect(Projects).toHaveLength(1);
  });

  test("Project List should be rendered correctly", () => {
    // act
    const wrapper = shallowMount(ProjectList, {
      localVue,
      vuetify,
      store,
      propsData: {
        listType: ProjectStatus.ACTIVE
      }
    });

    // assert
    expect(wrapper.element).toMatchSnapshot();
  });
});
