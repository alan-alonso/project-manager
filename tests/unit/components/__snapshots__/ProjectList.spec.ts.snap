// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectList.vue Project List should be rendered correctly 1`] = `
<v-container-stub
  class="lighten-1 mt-1 [object Object]"
  fluid="true"
  tag="div"
>
  <h2
    class="white--text text-center mb-2"
  >
    ACTIVE PROJECTS
  </h2>
   
</v-container-stub>
`;
